gensim/similarities/docsim.py

```diff
@@ -563,6 +564,107 @@ def __str__(self):
         return "%s<%i docs, %i features>" % (self.__class__.__name__, len(self), self.index.shape[1])
 
 
+class SoftCosineSimilarity(interfaces.SimilarityABC):
+    """Document similarity (like MatrixSimilarity) that uses Soft Cosine Similarity as a similarity
+    measure.
+
+    Parameters
+    ----------
+    corpus: list of lists of (int, float) two-tuples
+        A list of documents in the gensim document format.
+    similarity_matrix : scipy.sparse.csc_matrix
+        A term similarity matrix.
+    num_best : int
+        The number of results to retrieve for a query.
+
+    See Also
+    --------
+    gensim.models.keyedvectors.EuclideanKeyedVectors.similarity_matrix
+        A term similarity matrix produced from term embeddings.
+    gensim.matutils.softcossim
+        The Soft Cosine Similarity.
+
+    Examples
+    --------
+    >>> from gensim.models import Word2Vec
+    >>> from gensim.similarities import SoftCosineSimilarity
+    >>> from gensim.utils import simple_preprocess
+    >>> from gensim.corpora import Dictionary
+    >>> # Given a document collection "corpus", train a word2vec model.
+    >>> model = Word2Vec(corpus, workers=3, size=100)
+    >>> # Construct a bag-of-words corpus, a dictionary, and a term similarity matrix.
+    >>> dictionary = Dictionary(corpus)
+    >>> corpus = [dictionary.doc2bow(document) for document in corpus]
+    >>> similarity_matrix = model.wv.similarity_matrix(corpus, dictionary)
+    >>> index = SoftCosineSimilarity(corpus, similarity_matrix, num_best=10)
+    >>> # Make a query.
+    >>> query = 'Yummy! Great view of the Bellagio Fountain show.'
+    >>> sims = index[dictionary.doc2bow(simple_preprocess(query))]
+
+    See `Tutorial Notebook
+    <https://github.com/RaRe-Technologies/gensim/blob/develop/docs/notebooks/soft_cosine_tutorial.ipynb>`_
+    for more examples.
+    """
+
+    def __init__(self, corpus, similarity_matrix, num_best=None, chunksize=256):
+        self.corpus = corpus
+        self.num_best = num_best
+        self.chunksize = chunksize
+
+        # Normalization of features is undesirable, since soft cosine similarity requires special
+        # normalization using the similarity matrix. Therefore, we would just be normalizing twice,
+        # increasing the numerical error.
+        self.normalize = False
+
+        # index is simply an array from 0 to size of corpus.
+        self.index = numpy.array(range(len(corpus)))
```

Comment:
`numpy.arange(len(corpus))`

Reply:
Also, should corpus **always** have `__len__`? 

For example, if corpus something like this
```python
class MyCorpus(object):
    def __init__(self, fname):
        self.fname = fname
    def __iter__(self):
        with smart_open(self.fname) as infile:
            for line in infile:
                yield line
```

Reply:
This was taken straight from the `WmdSimilarity` class. Note that we also assume that a corpus has `__getitem__` (https://github.com/Witiko/gensim/blob/fc31b659f0d635c827e505b6accb17fd3f8b1925/gensim/similarities/docsim.py#L640).

Reply:
The first concern has been fixed in 03a967b580e8c5f45a5cbe12ebabf75064ab0d10.